### PR TITLE
Issue #157: Create new SQL Source

### PR DESCRIPTION
**Pull request by TNohaic**
Creation of new type of extension: SourceComputationExtension Creation of new SQL extension for SQL Sources

### CODE REVIEW

File: metricshub-sql-extension/src/main/java/org/sentrysoftware/metricshub/extension/sql/SqlClientExecutor.java [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+		for (SqlTable sqlTable : sqlTables) {
+			createAndInsert(sqlTable, connection);
+		}
+
+		final List<List<String>> result = new ArrayList<>();
+
+		try (Statement statement = connection.createStatement(); ResultSet resultSet = statement.executeQuery(query)) {
+			if (resultSet == null) {
+				log.error("the SQL query {} returned a null result.", query);
+				return null;
```
[the remark this answers is not in the document]
Comment (NassimBtk) [reply metricshub-sql-extension/src/main/java/org/sentrysoftware/metricshub/extension/sql/SqlClientExecutor.java:93]: The connection is not closed here, to make sure the connection is always closed, it is better to do something like this:

```java
		try (Connection connection = DriverManager.getConnection("jdbc:h2:mem:" + hostId + uuid)) {
			
			// Prepare the SQL tables
			for (SqlTable sqlTable : sqlTables) {
				createAndInsert(sqlTable, connection);
			}

			return executeQuery(query, connection);

		} catch (SQLException exception) {
			log.error("Error when creating the SQL database: {}", exception.getMessage());
			log.debug("SQL Exception: ", exception);
		}
```

[the remark this answers is not in the document]
Comment (NassimBtk) [reply metricshub-sql-extension/src/main/java/org/sentrysoftware/metricshub/extension/sql/SqlClientExecutor.java:93]: executeQuery:

```java
	/**
	 * Execute a SQL query on the connection object and return the result in a form
	 * of a list of lists of strings.
	 * 
	 * @param query      The query to execute.
	 * @param connection The connection to the database.
	 * @return The result of the query.
	 */
	private List<List<String>> executeQuery(final String query, Connection connection) {

		final List<List<String>> result = new ArrayList<>();

		try (Statement statement = connection.createStatement(); ResultSet resultSet = statement.executeQuery(query)) {

			if (resultSet == null) {
				log.error("The SQL query {} returned a null result.", query);
				return result;
			}

			final ResultSetMetaData rsmd = resultSet.getMetaData();

			final int columnsNumber = rsmd.getColumnCount();

			while (resultSet.next()) {
				final List<String> row = new ArrayList<>();
				for (int i = 1; i <= columnsNumber; i++) {
					row.add(resultSet.getString(i));
				}
				result.add(row);
			}
		} catch (SQLException exception) {
			log.error("Error when executing SQL query {}: {}", query, exception.getMessage());
			log.debug("SQL Exception: ", exception);
		}

		return result;
	}
	
```

File: metricshub-sql-extension/src/main/java/org/sentrysoftware/metricshub/extension/sql/SqlClientExecutor.java [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+			return;
+		}
+
+		try {
+			final Statement statement = connection.createStatement();
+
+			statement.execute(createTableQuery);
+			log.debug("Executing CREATE TABLE query: {}", createTableQuery);
+		} catch (SQLException exception) {
+			log.error("Error when executing query {}: {}", createTableQuery, exception);
```
Comment (NassimBtk) [inline metricshub-sql-extension/src/main/java/org/sentrysoftware/metricshub/extension/sql/SqlClientExecutor.java:134]:
```java
			log.error("Error when executing CREATE TABLE query {}: {}", createTableQuery, exception.getMessage());
			log.debug("CREATE TABLE SQL Exception: ", exception);
```

File: metricshub-sql-extension/src/main/java/org/sentrysoftware/metricshub/extension/sql/SqlClientExecutor.java [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+		final String insertTableQuery = insertTableQuery(sqlTable);
+		if (insertTableQuery == null) {
+			return;
+		}
+
+		try {
+			final Statement statement = connection.createStatement();
+
+			statement.execute(insertTableQuery);
+			log.trace("Executing INSERT TABLE query: {}", insertTableQuery);
```
Comment (NassimBtk) [inline metricshub-sql-extension/src/main/java/org/sentrysoftware/metricshub/extension/sql/SqlClientExecutor.java:147]: debug or trace ? The create table query is in debug level.

File: metricshub-sql-extension/src/main/java/org/sentrysoftware/metricshub/extension/sql/SqlClientExecutor.java [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+		} catch (SQLException exception) {
+			log.error("Error when executing query {}: {}", insertTableQuery, exception);
```
Comment (NassimBtk) [inline metricshub-sql-extension/src/main/java/org/sentrysoftware/metricshub/extension/sql/SqlClientExecutor.java:149]:
```java
			log.error("Error when executing INSERT TABLE query {}: {}", insertTableQuery, exception.getMessage());
			log.debug("INSERT TABLE SQL Exception: ", exception);
```

File: metricshub-sql-extension/src/main/java/org/sentrysoftware/metricshub/extension/sql/SqlClientExecutor.java [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+		}
+
+		final List<SqlColumn> sqlColumns = sqlTable.getColumns();
+
+		final List<String> columnValues = new ArrayList<>();
+
+		for (final List<String> row : table) {
+			final List<String> rowValues = new ArrayList<>();
+			for (final SqlColumn sqlColumn : sqlColumns) {
+				final String separator = sqlColumn.getType().contains("VARCHAR") ? "'" : "";
```
Comment (NassimBtk) [inline metricshub-sql-extension/src/main/java/org/sentrysoftware/metricshub/extension/sql/SqlClientExecutor.java:265]: For quoting, is it enough VARCHAR or are there other types such as TEXT or CHAR or maybe other types in H2?

File: metricshub-sql-extension/src/main/java/org/sentrysoftware/metricshub/extension/sql/SqlClientExecutor.java [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+				rowValues.add(separator + row.get(sqlColumn.getNumber() - 1) + separator);
```
Comment (NassimBtk) [inline metricshub-sql-extension/src/main/java/org/sentrysoftware/metricshub/extension/sql/SqlClientExecutor.java:266]: How `null` values are managed? should we insert SQL `NULL` ?

File: metricshub-sql-extension/src/main/java/org/sentrysoftware/metricshub/extension/sql/SqlClientExecutor.java (hunk 5 above)
Comment (NassimBtk) [inline metricshub-sql-extension/src/main/java/org/sentrysoftware/metricshub/extension/sql/SqlClientExecutor.java:265]: Getting  the value at the beginning is better `String value = row.get(sqlColumn.getNumber() - 1);`

File: metricshub-sql-extension/src/main/java/org/sentrysoftware/metricshub/extension/sql/SqlExtension.java [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+/**
+ * This class implements the {@link ISourceComputationExtension} contract, reports the supported features,
+ * processes Sql sources.
+ */
+@Slf4j
+public class SqlExtension implements ISourceComputationExtension {
+
+	@Override
+	public SourceTable processSource(Source source, String connectorId, TelemetryManager telemetryManager) {
+		final String hostname = telemetryManager.getHostConfiguration().getHostname();
```
Comment (NassimBtk) [inline metricshub-sql-extension/src/main/java/org/sentrysoftware/metricshub/extension/sql/SqlExtension.java:42]: `final String hostname = telemetryManager.getHostname();`

File: metricshub-sql-extension/src/main/java/org/sentrysoftware/metricshub/extension/sql/SqlExtension.java [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+			);
+			return SourceTable.empty();
+		}
+
+		final List<List<String>> executeSqlQuery = new SqlClientExecutor(telemetryManager, connectorId)
+			.executeQuery(sqlTables, query);
+
+		SourceTable sourceTable = new SourceTable();
+
+		if (executeSqlQuery != null) {
```
Comment (NassimBtk) [inline metricshub-sql-extension/src/main/java/org/sentrysoftware/metricshub/extension/sql/SqlExtension.java:79]: `.executeQuery(sqlTables, query);` shouldn't return null table please.

File: pom.xml [hunk 9]
```diff
@@ -342,6 +343,11 @@
 				<artifactId>ssh</artifactId>
 				<version>1.0.01</version>
 			</dependency>
+			<dependency>
```
Comment (NassimBtk) [inline pom.xml:346]: Remove this dependency please , it doesn't exist.

File: metricshub-sql-extension/pom.xml [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+		</dependency>
+		<dependency>
+			<groupId>org.snmp4j</groupId>
+			<artifactId>snmp4j-agent</artifactId>
+			<scope>test</scope>
+		</dependency>
+		<dependency>
+			<groupId>com.h2database</groupId>
+			<artifactId>h2</artifactId>
+			<version>2.2.224</version>
```
Comment (NassimBtk) [inline metricshub-sql-extension/pom.xml:86]: Remove the version it is managed in the main pom.

File: metricshub-engine/src/main/java/org/sentrysoftware/metricshub/engine/connector/model/common/SqlColumn.java [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+
+	@Override
+	public String toString() {
+		final StringBuilder stringBuilder = new StringBuilder();
+
+		stringBuilder.append("SQL Column: name=");
+		stringBuilder.append(name);
+		stringBuilder.append(", type=");
+		stringBuilder.append(type);
+		stringBuilder.append(", column number=");
```
[the remark this answers is not in the document]
Comment (TNohaic) [reply metricshub-engine/src/main/java/org/sentrysoftware/metricshub/engine/connector/model/common/SqlColumn.java:84]: That's how I set it at first, but I have found that it was not explicit at all when I was lloking at the logs.

[the remark this answers is not in the document]
Comment (NassimBtk) [reply metricshub-engine/src/main/java/org/sentrysoftware/metricshub/engine/connector/model/common/SqlColumn.java:84]: Not a big deal, keep it as it is 😉.

File: metricshub-engine/src/main/java/org/sentrysoftware/metricshub/engine/connector/model/common/SqlTable.java [hunk 12]
```diff
@@ -72,7 +74,7 @@ public class SqlTable implements Serializable {
 	 * @return A new instance of {@link SqlTable} with the same source, alias and columns.
 	 */
 	public SqlTable copy() {
-		return SqlTable.builder().source(source).alias(alias).columns(columns).build();
+		return SqlTable.builder().source(source).alias(alias).columns(new ArrayList<>(columns)).build();
```
Comment (NassimBtk) [inline metricshub-engine/src/main/java/org/sentrysoftware/metricshub/engine/connector/model/common/SqlTable.java:77]:
```java
	public SqlTable copy() {
		return SqlTable.builder().source(source).alias(alias).columns(columns.stream().map(SqlColumn::copy).collect(Collectors.toCollection(ArrayList::new))).build();
	}
```

File: metricshub-sql-extension/src/main/java/org/sentrysoftware/metricshub/extension/sql/SqlClientExecutor.java (hunk 5 above)
Comment (TNohaic) [reply metricshub-sql-extension/src/main/java/org/sentrysoftware/metricshub/extension/sql/SqlClientExecutor.java:277]: There is no TEXT type, but there is a CHARACTER type. I will update to `sqlColumn.getType().contains("CHAR")`

File: metricshub-sql-extension/src/main/java/org/sentrysoftware/metricshub/extension/sql/SqlClientExecutor.java (hunk 6 above)
Comment (TNohaic) [reply metricshub-sql-extension/src/main/java/org/sentrysoftware/metricshub/extension/sql/SqlClientExecutor.java]: I will add a unit test and change the code if needed.

**Review by NassimBtk — APPROVED** [review]

**Review by NassimBtk — APPROVED** [review]

**Pull request merged**